src/include/OpenImageIO/atomic.h

```diff
@@ -42,548 +42,72 @@
-    ///
-    T operator++ () { return fetch_add(1) + 1; }
-
-    /// Post-increment:  foo++
-    ///
-    T operator++ (int) {  return fetch_add(1); }
-
-    /// Pre-decrement:  --foo
-    ///
-    T operator-- () {  return fetch_sub(1) - 1; }
-
-    /// Post-decrement:  foo--
-    ///
-    T operator-- (int) {  return fetch_sub(1); }
-
-    /// Add to the value, return the new result
-    ///
-    T operator+= (T x) { return fetch_add(x) + x; }
-
-    /// Subtract from the value, return the new result
-    ///
-    T operator-= (T x) { return fetch_sub(x) - x; }
-
-    /// Logical and, return the new result
-    ///
-    T operator&= (T x) { return fetch_and(x) & x; }
-
-    /// Logical or, return the new result
-    ///
-    T operator|= (T x) { return fetch_or(x) | x; }
-
-    /// Logical xor, return the new result
-    ///
-    T operator^= (T x) { return fetch_xor(x) ^ x; }
-
-    bool bool_compare_and_swap (T compareval, T newval) {
-        return atomic_compare_and_exchange (&m_val, compareval, newval);
-    }
-
-    bool compare_exchange_weak (T& compareval, T newval,
-                                memory_order order = memory_order_seq_cst) {
-        return atomic_compare_and_exchange (&m_val, compareval, newval,
-                                            true, order, order);
-    }
-    bool compare_exchange_strong (T& compareval, T newval,
-                                  memory_order order = memory_order_seq_cst) {
-        return atomic_compare_and_exchange (&m_val, compareval, newval,
-                                            false, order, order);
-    }
-
-    T operator= (const atomic &x) {
-        T r = x();
-        *this = r;
-        return r;
-    }
-
-private:
-#ifdef __arm__
-    OIIO_ALIGN(8)
-#endif
-    volatile mutable T m_val;
-
-    // Disallow copy construction by making private and unimplemented.
-    atomic (atomic const &);
-};
-
-#endif
-
-
-
 typedef atomic<int> atomic_int;
 typedef atomic<long long> atomic_ll;
 
 
 
+/// Atomically set avar to the minimum of its current value and bval.
 template<typename T>
 OIIO_FORCEINLINE void
 atomic_min (atomic<T> &avar, const T &bval)
 {
     do {
         T a = avar.load();
-        if (a <= bval || avar.compare_exchange_strong (a,bval))
+        if (a <= bval || avar.compare_exchange_weak (a,bval))
             break;
-    } while (1);
+    } while (true);
 }
 
 
+/// Atomically set avar to the maximum of its current value and bval.
 template<typename T>
 OIIO_FORCEINLINE void
 atomic_max (atomic<T> &avar, const T &bval)
 {
     do {
         T a = avar.load();
-        if (a >= bval || avar.compare_exchange_strong (a,bval))
+        if (a >= bval || avar.compare_exchange_weak (a,bval))
             break;
-    } while (1);
+    } while (true);
+}
+
+
+
+// Add atomically to a float and return the original value.
+OIIO_FORCEINLINE float
+atomic_fetch_add (float *af, float f)
+{
+    // C++11 doesn't support an atomic<float> with math ops.
```

Comment:
Updated. Like this?

I had done the `float*` mostly because I wasn't sure whether to trust compare_and_swap for `std::atomic<float>`, but I've confirmed that it works and has identical performance to my int trick.